### ISSUE TITLE
Dev build uses ExternalAssetHandler, but production build does not?

### ISSUE THREAD

**Issue by mholt** [wailsapp/wails#2109]
### Description

When I run `wails dev` and my frontend navigates to `/gallery/`, my external assets handler rewrites the request to serve `/gallery/index.html`. However, with the exact same code compiled with `wails build`, the exact same page load yields 500 Internal Server Error.

### To Reproduce

Here's my code:

```
func (a *App) Run() error {
	return wails.Run(&options.App{
		Title:     "Timelinize",
		Width:     1200,
		Height:    800,
		Frameless: true,
		AssetServer: &assetserver.Options{
			Assets:     a.frontend,
			Middleware: a.StaticWebsiteMiddleware,
		},
		OnStartup: func(ctx context.Context) { a.ctx = ctx },
		Bind:      []any{a},
	})
}
```

Notice I set up a custom AssetServer with my own Middleware. The Middleware basically determines whether the request needs to be rewritten or have server-side templates evaluated, and if not, it passes the request to the next handler (wails).

Anyway, this all works great with `wails dev`:

```
DEB | [ExternalAssetHandler] Loading 'http://localhost:3000/gallery/'
```

But running the same program from `wails build` produces:

```
ERR | [AssetHandler] Unable to load file 'gallery/': open gallery/: invalid argument
```

It seems to me that the production build is not using ExternalAssetHandler.

### Expected behaviour

I'd expect `wails dev` and `wails build` programs to run the same in this regard.

### Screenshots

(Probably not applicable, since it's text logs. The UI just crashes with "Internal server error" 

### Attempted Fixes

I'm not sure what to try yet. This is a new feature, so I'm not surprised if I'm either using it wrong or there's a little bug. :bug: :slightly_smiling_face: 

### System Details

```shell
Wails CLI v2.2.0

Scanning system - Please wait (this may take a long time)...Done.

System
------
OS:             Pop!_OS
Version:        22.04
ID:             pop
Go Version:     go1.19.1
Platform:       linux
Architecture:   amd64

Wails
------
Version:                v2.2.0
Package Manager:        apt

Dependency      Package Name            Status          Version
----------      ------------            ------          -------
*docker         docker.io               Available       20.10.12-0ubuntu4
gcc             build-essential         Installed       11.3.0
libgtk-3        libgtk-3-dev            Installed       3.24.33-1ubuntu2
libwebkit       libwebkit2gtk-4.0-dev   Installed       2.38.2-0ubuntu0.22.04.2
npm             npm                     Installed       8.5.1~ds-1
*nsis           nsis                    Available       3.08-2
pkg-config      pkg-config              Installed       0.29.2

* - Optional Dependency

Diagnosis
---------
Your system is ready for Wails development!
Optional package(s) installation details: 
  - docker: sudo apt install docker.io
  - nsis: sudo apt install nsis
```

### Additional context

Thank you again for Wails, it's awesome!

**Comment by stffabi** [comment]
Thanks for bringing this up @mholt.

The `ExternalAssetHandler` is something very specialised only for the dev mode. It handles all the integration with the external dev server and proxies requests over to it. So the final destination is a full featured web server and can handle a request to `/gallery/` by serving the `index.html` there.

When using the production build, there's an internal 'dummy' WebServer which serves the content. As such it has some missing features, but step by step we are adding more to it. Recently we added support for range requests and this issue is also a missing functionality.

I already have a branch ready for this, but don't know if I will finish it today.

**Comment by mholt** [comment]
Oh, thanks. I didn't realize it was only for dev mode. I would definitely expect that my dev mode asset server is the same as for the production builds. My program won't work if it can't also do rewriting and template evaluation for the frontend files.

**Comment by stffabi** [comment]
> Oh, thanks. I didn't realize it was only for dev mode. I would definitely expect that my dev mode asset server is the same as for the production builds

Unfortunately that's not possible to have them 100% identical at least not without starting a webserver on localhost in production mode and bundle e.g. Vite and start it in production. But not having to start a webserver is a main feature of Wails. There are cases where a webserver on localhost might not be a problem, but e.g. for commercial products this might be problematic. A company might not like to have the whole frontend and the full backend-frontend communication being exposed.

Maybe once we could write a technical whitepaper with all the technical details. More people asked about this lately.

>My program won't work if it can't also do rewriting and template evaluation for the frontend files.

The middleware can also be used in production. You can also rewrite the response body there like in dev mode. But there are limitations of what can be used from the request and response, see the feature matrix in [AssetServer](https://wails.io/docs/next/reference/options#assetserver). What do you mean by rewriting? Do you mean http.Redirects?

**Comment by mholt** [comment]
Ah, bummer. I see what you mean, though, that's a good point.

This is the code for my middleware handler:

```go 
func (a *App) serveFrontend(w http.ResponseWriter, r *http.Request, next http.Handler) {
	// don't do any actual handling yet; just set up the request middleware stuff, logging, etc...
	start := time.Now()

	// shouldBuf determines whether to execute templates on this response,
	// since generally we will not want to execute for images, etc.
	allowTemplates := true
	shouldBuf := func(status int, header http.Header) bool {
		if !allowTemplates {
			return false
		}
		ct := header.Get("Content-Type")
		for _, mt := range []string{"text/"} {
			if strings.Contains(ct, mt) {
				return true
			}
		}
		return false
	}

	buf := bufPool.Get().(*bytes.Buffer)
	buf.Reset()
	defer bufPool.Put(buf)

	rec := caddyhttp.NewResponseRecorder(w, buf, shouldBuf)

	var err error

	defer func() {
		logFn := a.log.Info
		if err != nil {
			logFn = a.log.Error
		}
		// the log message is intentionally specific to bust log sampling here
		logFn(r.Method+" "+r.RequestURI,
			zap.String("method", r.Method),
			zap.String("uri", r.RequestURI),
			zap.Int("status", rec.Status()),
			zap.Int("size", rec.Size()),
			zap.Duration("duration", time.Since(start)),
			zap.Error(err))
	}()

	// ok, we're all set up, so actual handling can go below here

	// "sir, this is an Arby's"
	if r.Method != http.MethodGet && r.Method != http.MethodHead {
		next.ServeHTTP(rec, r)
		return
	}

	// serve special data files
	if strings.HasPrefix(r.URL.Path, "/foo/") || strings.HasPrefix(r.URL.Path, "/bar/") {
		parts := strings.SplitN(r.URL.Path, "/", 6)
		if len(parts) < 4 {
			http.Error(rec, "insufficient path", http.StatusBadRequest)
			return
		}
		repoID := parts[2]
		dataOrItem := parts[3]

		// ... application logic ...

		switch dataOrItem {
		case "data", "assets":
			// data or asset file; not in frontend assets
			allowTemplates = false // don't evaluate templates on data files
			tl.fileServer.ServeHTTP(rec, r)
			return
		case "item":
			if len(parts) > 5 {
				http.Error(w, "invalid item path", http.StatusBadRequest)
				return
			}
			r.URL.Path = "/item"
		}
	}

	// even though we're serving HTML files, we reject ugly URLs containing
	// the .html file extension (or "index") since they're user-facing
	if strings.HasSuffix(r.URL.Path, ".html") || strings.HasSuffix(r.URL.Path, "/index") {
		http.Error(rec, "ugly URLs verboten", http.StatusBadRequest)
		return
	}

	if r.URL.Path == "/" {
		r.URL.Path = "index.html"
	}

	// we enforce clean URLs but an upstream handler serves static content,
	// so append the file extension implicitly if not a directory
	if !strings.HasSuffix(r.URL.Path, ".html") {
		_, err := fs.Stat(a.frontend, strings.TrimPrefix(r.URL.Path, "/")+".html")
		if err == nil {
			r.URL.Path += ".html"
		}
	}

	if path.Ext(r.URL.Path) == ".html" {
		// ... application logic ...

		if setupNeeded && r.URL.Path != "/setup.html" {
			to := "/setup"
			r.URL.Path = strings.TrimSuffix(r.URL.Path, "index.html")
			if uri := r.URL.RequestURI(); uri != "/" {
				// make sure we return to where they wanted to go
				v := make(url.Values)
				v.Set("return", uri)
				to += "?" + v.Encode()
			}
			redirect(w, to)
			return
		} else if !setupNeeded && r.URL.Path == "/setup.html" {
			redirect(w, "/")
			return
		}
	}

	// for clean URLs and methods we recognize, capture the static file response
	next.ServeHTTP(rec, r)
	if !rec.Buffered() {
		return
	}

	err = executeTemplate(r, a.frontend, rec.Buffer())
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	rec.Header().Set("Content-Length", strconv.Itoa(buf.Len()))
	rec.Header().Del("Accept-Ranges") // we don't know ranges for dynamically-created content
	rec.Header().Del("Last-Modified") // useless for dynamic content since it's always changing

	// we don't know a way to quickly generate etag for dynamic content,
	// and weak etags still cause browsers to rely on it even after a
	// refresh, so disable them until we find a better way to do this
	rec.Header().Del("Etag")

	rec.WriteResponse()
}
```

I'm mostly doing internal rewriting, but I manipulate headers and also do redirects too.

**Comment by stffabi** [comment]
Draft PR #2110 is up which should solve this issue.

I think most of the middleware should works, apart from the `http.redirects` on Linux. This would need support for newer WebKitGTK 2.36+ see #1498

**Comment by leaanthony** [comment]
Nice one @stffabi! Looks like 2.36+ would help on a number of fronts. I'll see if that's something I can look at this week. Perhaps #2044 could help for fallback functionality if it's not there. I'm hoping it's not an all or nothing thing...

**Comment by stffabi** [comment]
That would be awesome @leaanthony.

Just a side note, I've just tested the http redirects, seems like they work on Windows but on macOS they don't work on a custom scheme 😕

**Comment by stffabi** [comment]
I can confirm that http redirects code 307, 308 don't work with Webkit webviews. So this affects macOS and Linux.

**Comment by leaanthony** [comment]
Thanks for confirming!

**Comment by stffabi** [comment]
Seems like this commit added the RedirectionAPI for cocoa but it's still not available in macOS. And the API is also missing for WebKit2GTK. According to that commit we would need to handle http Redirects in a special way by calling another API to finish the request.

**Comment by stffabi** [comment]
I think we could still make it possible to serve the index.html if a directory is requested that ends with a slash in the url.
Basically the referenced PR already does this. We only can't do a redirect if it's a directory but the url does not end with a slash, but since one is in control of the frontend assets one could easily just make sure there's the slash.

Nevertheless redirecting support would be awesome, and that's currently blocked by apple and webkit2gtk. But I think we could track that in a separate issue.

**Comment by leaanthony** [comment]
> But not having to start a webserver is a main feature of Wails. There are cases where a webserver on localhost might not be a problem, but e.g. for commercial products this might be problematic. A company might not like to have the whole frontend and the full backend-frontend communication being exposed.

Someone has opened a ticket for this exact feature.

**Comment by leaanthony** [comment]
I don't even know if it's possible but I wonder if we could write a `fetch` compatible that mimicked the unsupported methods via query parameters and then rewrite it all server side?

**Comment by stffabi** [comment]
> I don't even know if it's possible but I wonder if we could write a `fetch` compatible that mimicked the unsupported methods via query parameters and then rewrite it all server side?

What unsupported methods to you mean?

**Comment by stffabi** [comment]
> > I don't even know if it's possible but I wonder if we could write a `fetch` compatible that mimicked the unsupported methods via query parameters and then rewrite it all server side?
> 
> What unsupported methods to you mean?

WebKit2GTK 2.36+ supports all http methods apart from request bodies and http redirects with 30x status codes.

**Comment by leaanthony** [comment]
Those ones I guess 😉 Ok, we can spend the effort supporting 2.36 instead 👍

**Comment by stffabi** [comment]
> Those ones I guess 😉 Ok, we can spend the effort supporting 2.36 instead 👍

Already on it and almost finished 😀
